go/vt/discovery/healthcheck.go

```diff
@@ -458,28 +471,64 @@ func (hc *HealthCheckImpl) checkConn(hcc *healthCheckConn, name string) {
 	defer hc.finalizeConn(hcc)
 
 	// Initial notification for downstream about the tablet existence.
-	hcc.mu.Lock()
-	ts := hcc.tabletStats
-	hcc.mu.Unlock()
-	if hc.listener != nil {
-		hc.listener.StatsUpdate(&ts)
-	}
+	hc.updateHealth(hcc.tabletStats.Copy(), hcc.conn)
 	hc.initialUpdatesWG.Done()
 
 	retryDelay := hc.retryDelay
 	for {
-		ctx, cancel := context.WithCancel(hcc.ctx)
-		hcc.mu.Lock()
-		hcc.streamCancelFunc = cancel
-		hcc.mu.Unlock()
+		streamCtx, streamCancel := context.WithCancel(hcc.ctx)
+
+		// Setup a watcher that restarts the timer every time an update is received.
+		// If a timeout occurs for a serving tablet, we make it non-serving and send
+		// a status update. The stream is also terminated so it can be retried.
+		// servingStatus feeds into the serving var, which keeps track of the serving
+		// status transmitted by the tablet.
+		servingStatus := make(chan bool, 5)
+		serving := hcc.tabletStats.Serving
+		timedout := false
+		go func() {
+			for {
+				select {
+				case serving = <-servingStatus:
+					continue
+				case <-time.After(hc.healthCheckTimeout):
+					// Ignore if not serving.
+					if !serving {
+						continue
+					}
+					timedout = true
+					streamCancel()
+					return
+				case <-streamCtx.Done():
+					// If stream returns while serving is false, the function
+					// will get stuck in an infinite loop. This code path
+					// breaks the loop.
+					return
+				}
+			}
+		}()
 
 		// Read stream health responses.
-		hcc.stream(ctx, hc, func(shr *querypb.StreamHealthResponse) error {
+		hcc.stream(streamCtx, hc, func(shr *querypb.StreamHealthResponse) error {
 			// We received a message. Reset the back-off.
 			retryDelay = hc.retryDelay
+			servingStatus <- shr.Serving
```

Comment:
Although very very unlikely in practice, it seems like bad things would happen if the watcher goroutine got stalled and the `servingStatus` channel does actually fill up. Instead of 5 slots would it make more sense to just have one and then select here in the unlikely event it's blocked?